VisualRecognitionApi/src/main/java/javax/visrec/spi/ServiceProvider.java

```diff
@@ -0,0 +1,126 @@
+public abstract class ServiceProvider {
+
+    /** The lock to populate and mutate the providers list. */
+    private static final Object LOCK = new Object();
+
+    private static List<ServiceProvider> providers;
+
+    protected ServiceProvider() {
+        // Prevent instantiation, only allowed by subclasses.
+    }
+
+    /**
+     * If multiple implementations of the {@link ServiceProvider} are found on the classpath, then the
+     * {@link ServiceProvider} with the highest value of priority will be used by default.
+     * @return The priority (default = 0(
+     */
+    public int getPriority() {
+        return 0;
+    }
+
+    /**
+     * Get the {@link BuilderService}
+     * @return builder service.
+     */
+    public abstract BuilderService getBuilderService();
+
+    /**
+     * Get the {@link ClassifierService}
+     * @return classifier service
+     */
+    public abstract ClassifierService getClassifierService();
+
+    /**
+     * Get the current {@link ServiceProvider}
+     * @return service provider.
+     * @throws IllegalStateException If there are no service providers found.
+     */
+    public static ServiceProvider current() {
+        final ServiceProvider p = getProviders().get(0);
+        if (Objects.isNull(p)) {
+            throw new IllegalStateException("No ServiceProvider found");
+        }
+        return p;
+    }
+
+    /**
+     * Set the current {@link ServiceProvider}
+     * @param provider The {@link ServiceProvider} to be set as current.
+     * @throws IllegalStateException If there are no service providers found.
+     * @throws IllegalArgumentException If the {@link ServiceProvider} given by the parameters is not known
+     * in the existing list of providers.
+     */
+    public static void setCurrent(final ServiceProvider provider) {
+        Objects.requireNonNull(provider);
+
+        synchronized (LOCK) {
+            final List<ServiceProvider> foundProviders = getProviders();
+            if (foundProviders.isEmpty()) {
+                throw new IllegalStateException("No providers found.");
+            }
+            if (!foundProviders.contains(provider)) {
+                throw new IllegalArgumentException("ServiceProvider given through the parameters is not known.");
+            }
+
+            // Copying list, removes the provider from the arguments and prepends it upfront
+            // on the copied list.
+            final List<ServiceProvider> copiedProviders = new LinkedList<>();
+            Collections.copy(copiedProviders, foundProviders);
+            copiedProviders.remove(provider);
+            copiedProviders.add(0, provider);
+
+            // Make the list unmodifiable to prevent illegal modification
+            providers = Collections.unmodifiableList(copiedProviders);
+        }
+    }
+
+    /**
+     * Gets a list of all available {@link ServiceProvider}s.
+     * @return service providers.
+     */
+    public static List<ServiceProvider> available() {
```

Comment:
Not sure if this method provides any special purpose. It is not doing anything more than getProviders, it can be removed I guess.

Reply:
I'll remove this method and rename getProviders() to available().

Reply:
The `ServiceProvider` may be accessed by any developer using the API, it doesn't mean they would. The JavaDoc is clear enough to understand the functionality of the method. We could rename the method to `availableProviders()` but then I'd also like to rename `current()` to `currentProvider()` to keep it consistent. I'm open to other suggestions regarding the method names :) !

Reply:
Fair enough.